### PR TITLE
Add new hook for add-order-item

### PR DESCRIPTION
**Pull request by pinalj**
Added a new hook for validations in the add_order_item() called in
WC->Orders->Add/Edit Orders.

Allow plugins to perform validations/changes to the order before a product is added to the order.

### CODE REVIEW

**Review by claudiosanches — CHANGES_REQUESTED** [review]
Thanks for your help! Just a minor change before merge.

File: includes/class-wc-ajax.php [hunk 1]
```diff
@@ -898,7 +898,11 @@ public static function add_order_item() {
 				if ( ! $product ) {
 					throw new Exception( __( 'Invalid product ID', 'woocommerce' ) . ' ' . $product_id );
 				}
+				$check = apply_filters( 'woocommerce_add_item_check', $product, $order, $qty );
```
[the remark this answers is not in the document]
Comment (pinalj) [reply includes/class-wc-ajax.php]: Thanks, updated the hook name as suggested.

**Review by claudiosanches — APPROVED** [review]
Thanks for your contribution!

**Comment by TimothyBJacobs** [comment]
@claudiosanches This doesn't seem right, if a plugin filters to return a `WP_Error`, any later filters won't be able to determine what the product is. Shouldn't there be a separate first parameter for the `WP_Error` check?

**Comment by claudiosanches** [comment]
@TimothyBJacobs no need for an extra parameter, also this will stop immediately.

**Comment by TimothyBJacobs** [comment]
I'm not seeing that.

The filter is `$check = apply_filters( 'woocommerce_ajax_add_order_item_validation', $product, $order, $qty );`.

If I have two plugins, and the first does.

```
add_filter( 'woocommerce_ajax_add_order_item_validation', function( $product_or_error, $order ) {
    if ( ! isset( $_POST['some_key'] ) && $product->get_meta() ) {
        return new WP_Error();
    }

    return $product_or_error;
}, 10, 2 );
```

```
add_filter( 'woocommerce_ajax_add_order_item_validation', function( $product_or_error, $order ) {
    if ( ! isset( $_POST['some_key'] ) && $product_or_error->get_meta() ) {
        return new WP_Error();
    }

    return $product_or_error;
}, 10, 2 );
```

Won't this now fatal because `$product_or_error` is a `WP_Error` instance from the first filter that applies?

**Comment by claudiosanches** [comment]
@TimothyBJacobs you need to check if is an `WP_Error` instance instead of a `WP_Product` instance.

**Comment by TimothyBJacobs** [comment]
Yes, that will avoid the fatal, but now I no longer have access to the product. And if I am validating based on some product attribute, I won't be able to.

Adding a separate parameter is how WC works elsewhere. For instance:

```
$validation_error  = new WP_Error();
$validation_error  = apply_filters( 'woocommerce_process_registration_errors', $validation_error, $username, $password, $email );
```

```
$validation_error = new WP_Error();
$validation_error = apply_filters( 'woocommerce_process_login_errors', $validation_error, $creds['user_login'], $creds['user_password'] );
```

```
$passed_validation = apply_filters( 'woocommerce_add_to_cart_validation', true, $item, $quantity );
```

```
$passed_validation = apply_filters( 'woocommerce_update_cart_validation', true, $cart_item_key, $values, $quantity );
```

**Comment by claudiosanches** [comment]
@TimothyBJacobs please feel free to submit a Pull Request, if you are so strongly against this filter, we can change, no problem.

**Pull request merged**